File: src/engine/scoring.js
import { ruleSets } from "./rules.js";
import { countComboTag } from "./combos.js";

/* ============================================================================
 * Scoring engine
 * - base score and payout multiplier
 * - bak detection
 * - pi/gukjin helper utilities
 * ========================================================================== */

const NON_BRIGHT_KWANG_ID = "L0";
const GUKJIN_CARD_ID = "I0";

/* 1) Shared card helpers */
function uniqueCards(cards = []) {
  const seen = new Set();
  return cards.filter((c) => {
    if (!c || seen.has(c.id)) return false;
    seen.add(c.id);
    return true;
  });
}

export function calculateScore(player, opponent, ruleKey) {
  const rules = ruleSets[ruleKey];
  const baseInfo = calculateBaseScore(player, rules);
  const goBonus = player.goCount;

  // base/total is base score + GO bonus; payoutTotal applies multipliers.
  const pointTotal = baseInfo.base + goBonus;

  let payoutMultiplier = 1;
  if (player.goCount >= 3) payoutMultiplier *= 2 ** (player.goCount - 2);
  if (player.events.shaking > 0) payoutMultiplier *= 2 ** player.events.shaking;
  if (player.events.bomb > 0) payoutMultiplier *= 2 ** player.events.bomb;

  const bak = detectBak(player, opponent, ruleKey);
  payoutMultiplier *= bak.multiplier;

  return {
    base: pointTotal,
    multiplier: payoutMultiplier,
    total: pointTotal,
    payoutTotal: pointTotal * payoutMultiplier,
    bak,
    breakdown: { ...baseInfo.breakdown, goBonus }
  };
}


/* 2) Base-score components */
export function calculateBaseScore(player, rules = ruleSets.A) {
  const kwang = uniqueCards(player.captured.kwang || []);
  const ribbon = uniqueCards(player.captured.ribbon || []);
  const fiveCards = scoringFiveCards(player);
  const fiveCount = fiveCards.length;
  const ribbonCount = ribbon.length;
  const junkCount = scoringPiCount(player);

  const kwangBase = kwangBaseScore(kwang);
  const fiveBase = fiveCount >= 5 ? fiveCount - 4 : 0;
  const ribbonBase = ribbonCount >= 5 ? ribbonCount - 4 : 0;
  const junkBase = junkCount >= 10 ? junkCount - 9 : 0;
  const ribbonSetBonus = ribbonBonus(ribbon);
  const fiveSetBonus = fiveBonus(fiveCards);
  const base = kwangBase + fiveBase + ribbonBase + junkBase + ribbonSetBonus + fiveSetBonus;

  return {
    base,
    breakdown: {
      kwangBase,
      fiveBase,
      ribbonBase,
      junkBase,
      ribbonSetBonus,
      fiveSetBonus,
      piCount: junkCount,
      gukjinMode: getGukjinMode(player)
    }
  };
}

function ribbonBonus(ribbons) {
  const hasRed = countComboTag(ribbons, "redRibbons") >= 3;
  const hasBlue = countComboTag(ribbons, "blueRibbons") >= 3;
  const hasPlant = countComboTag(ribbons, "plainRibbons") >= 3;
  let bonus = 0;
  if (hasRed) bonus += 3;
  if (hasBlue) bonus += 3;
  if (hasPlant) bonus += 3;
  return bonus;
}

function fiveBonus(fives) {
  return countComboTag(fives, "fiveBirds") >= 3 ? 5 : 0;
}

function kwangBaseScore(kwangCards) {
  const count = kwangCards.length;
  if (count < 3) return 0;
  if (count === 3) return kwangCards.some((c) => c?.id === NON_BRIGHT_KWANG_ID) ? 2 : 3;
  if (count === 4) return 4;
  return 15;
}

/* 3) Bak detection */
function detectBak(player, opponent, ruleKey) {
  const rules = ruleSets[ruleKey];
  const opponentPi = scoringPiCount(opponent);
  const playerPi = scoringPiCount(player);
  const playerFiveCount = scoringFiveCards(player).length;
  const opponentFiveCount = scoringFiveCards(opponent).length;

  const bak = {
    gwang:
      uniqueCards(player.captured.kwang || []).length >= 3 &&
      uniqueCards(opponent.captured.kwang || []).length === 0,
    // ?쇰컯? 怨??щ?? 臾닿?. ?? ?곷? ?쇨? 0?대㈃ 硫대컯(?쇰컯 硫댁젣).
    pi: opponentPi >= 1 && opponentPi <= 7 && playerPi >= 10,
    // User rule: mongBak is checked on win condition context without GO prerequisite.
    mongBak: opponentFiveCount === 0 && playerFiveCount >= 7
  };

  let multiplier = 1;
  if (bak.gwang) multiplier *= rules.bakMultipliers.gwang;
  if (bak.pi) multiplier *= rules.bakMultipliers.pi;
  if (bak.mongBak) multiplier *= rules.bakMultipliers.mongBak;
  return { ...bak, multiplier };
}

/* 4) Pi/gukjin helpers */
export function piValue(card) {
  const explicit = Number(card?.piValue);
  if (Number.isFinite(explicit) && explicit > 0) return explicit;
  return 1;
}

export function sumPiValues(cards = []) {
  return uniqueCards(cards).reduce((sum, card) => sum + piValue(card), 0);
}

export function isGukjinCard(card) {
  return card?.id === GUKJIN_CARD_ID;
}

export function getGukjinMode(player) {
  return player?.gukjinMode === "junk" ? "junk" : "five";
}

export function scoringFiveCards(player) {
  const fives = uniqueCards(player.captured.five || []);
  if (getGukjinMode(player) === "five") return fives;
  return fives.filter((c) => !(isGukjinCard(c) && !c.gukjinTransformed));
}

export function scoringPiCount(player) {
  const basePi = sumPiValues(player.captured.junk || []);
  if (getGukjinMode(player) !== "junk") return basePi;
  const hasGukjin = uniqueCards(player.captured.five || []).some((c) => isGukjinCard(c) && !c.gukjinTransformed);
  return hasGukjin ? basePi + 2 : basePi;
}

File: experiments/k_hyperneat_matgo/py_eval/scoring.py
from __future__ import annotations

from .combos import count_combo_tag, unique_cards
from .models import Card, PlayerState
from .rules import RULE_SETS, RuleSet, normalize_rule_key


NON_BRIGHT_KWANG_ID = "L0"
GUKJIN_CARD_ID = "I0"


def calculate_score(player: PlayerState, opponent: PlayerState, rule_key: str | None = None) -> dict[str, object]:
    rules = RULE_SETS[normalize_rule_key(rule_key)]
    base_info = calculate_base_score(player, rules)
    go_bonus = int(player.go_count or 0)
    point_total = int(base_info["base"]) + go_bonus

    payout_multiplier = 1
    if player.go_count >= 3:
        payout_multiplier *= 2 ** (player.go_count - 2)
    if player.events.shaking > 0:
        payout_multiplier *= 2 ** player.events.shaking
    if player.events.bomb > 0:
        payout_multiplier *= 2 ** player.events.bomb

    bak = detect_bak(player, opponent, rule_key)
    payout_multiplier *= int(bak["multiplier"])

    breakdown = dict(base_info["breakdown"])
    breakdown["goBonus"] = go_bonus
    return {
        "base": point_total,
        "multiplier": payout_multiplier,
        "total": point_total,
        "payoutTotal": point_total * payout_multiplier,
        "bak": bak,
        "breakdown": breakdown,
    }


def calculate_base_score(player: PlayerState, rules: RuleSet | None = None) -> dict[str, object]:
    active_rules = rules or RULE_SETS["A"]
    kwang = unique_cards(player.captured.kwang)
    ribbons = unique_cards(player.captured.ribbon)
    five_cards = scoring_five_cards(player)
    five_count = len(five_cards)
    ribbon_count = len(ribbons)
    junk_count = scoring_pi_count(player)

    kwang_base = kwang_base_score(kwang)
    five_base = five_count - 4 if five_count >= 5 else 0
    ribbon_base = ribbon_count - 4 if ribbon_count >= 5 else 0
    junk_base = junk_count - 9 if junk_count >= 10 else 0
    ribbon_set_bonus = ribbon_bonus(ribbons)
    five_set_bonus = five_bonus(five_cards)
    base = kwang_base + five_base + ribbon_base + junk_base + ribbon_set_bonus + five_set_bonus

    return {
        "base": base,
        "breakdown": {
            "kwangBase": kwang_base,
            "fiveBase": five_base,
            "ribbonBase": ribbon_base,
            "junkBase": junk_base,
            "ribbonSetBonus": ribbon_set_bonus,
            "fiveSetBonus": five_set_bonus,
            "piCount": junk_count,
            "gukjinMode": get_gukjin_mode(player),
            "goMinScore": active_rules.go_min_score,
        },
    }


def ribbon_bonus(ribbons: list[Card]) -> int:
    bonus = 0
    if count_combo_tag(ribbons, "redRibbons") >= 3:
        bonus += 3
    if count_combo_tag(ribbons, "blueRibbons") >= 3:
        bonus += 3
    if count_combo_tag(ribbons, "plainRibbons") >= 3:
        bonus += 3
    return bonus


def five_bonus(fives: list[Card]) -> int:
    return 5 if count_combo_tag(fives, "fiveBirds") >= 3 else 0


def kwang_base_score(kwang_cards: list[Card]) -> int:
    count = len(kwang_cards)
    if count < 3:
        return 0
    if count == 3:
        return 2 if any(card.id == NON_BRIGHT_KWANG_ID for card in kwang_cards) else 3
    if count == 4:
        return 4
    return 15


def detect_bak(player: PlayerState, opponent: PlayerState, rule_key: str | None = None) -> dict[str, object]:
    rules = RULE_SETS[normalize_rule_key(rule_key)]
    opponent_pi = scoring_pi_count(opponent)
    player_pi = scoring_pi_count(player)
    player_five_count = len(scoring_five_cards(player))
    opponent_five_count = len(scoring_five_cards(opponent))

    gwang = len(unique_cards(player.captured.kwang)) >= 3 and len(unique_cards(opponent.captured.kwang)) == 0
    pi = opponent_pi >= 1 and opponent_pi <= 7 and player_pi >= 10
    mong_bak = opponent_five_count == 0 and player_five_count >= 7

    multiplier = 1
    if gwang:
        multiplier *= rules.bak_multipliers["gwang"]
    if pi:
        multiplier *= rules.bak_multipliers["pi"]
    if mong_bak:
        multiplier *= rules.bak_multipliers["mongBak"]
    return {"gwang": gwang, "pi": pi, "mongBak": mong_bak, "multiplier": multiplier}


def pi_value(card: Card | None) -> int:
    if not card:
        return 1
    explicit = int(card.pi_value or 0)
    return explicit if explicit > 0 else 1


def sum_pi_values(cards: list[Card] | None = None) -> int:
    return sum(pi_value(card) for card in unique_cards(cards))


def is_gukjin_card(card: Card | None) -> bool:
    return bool(card and card.id == GUKJIN_CARD_ID)


def get_gukjin_mode(player: PlayerState) -> str:
    return "junk" if player.gukjin_mode == "junk" else "five"


def scoring_five_cards(player: PlayerState) -> list[Card]:
    fives = unique_cards(player.captured.five)
    if get_gukjin_mode(player) == "five":
        return fives
    return [card for card in fives if not is_gukjin_card(card)]


def scoring_pi_count(player: PlayerState) -> int:
    base_pi = sum_pi_values(player.captured.junk)
    if get_gukjin_mode(player) != "junk":
        return base_pi
    has_gukjin = any(is_gukjin_card(card) for card in unique_cards(player.captured.five))
    return base_pi + 2 if has_gukjin else base_pi
